napari/utils/notifications.py

```diff
@@ -0,0 +1,220 @@
+        self,
+        message: str,
+        severity: Union[
+            str, NotificationSeverity
+        ] = NotificationSeverity.WARNING,
+        # source: Optional[str] = None, # TODO
+        actions: ActionSequence = (),
+        type: str = 'notification',
+        native: Any = None,
+        **kwargs,
+    ):
+
+        super().__init__(type, **kwargs)
+        self.message = message
+        self.severity = NotificationSeverity(severity)
+        # self.source = source  # TODO
+        self.actions = actions
+
+        # let's store when the object was created;
+        self.date = datetime.now()
+
+    @classmethod
+    def from_exception(cls, exc: BaseException, **kwargs) -> 'Notification':
+        return ErrorNotification(exc, **kwargs)
+
+    @classmethod
+    def from_warning(cls, warning: Warning, **kwargs) -> 'Notification':
+        return WarningNotification(warning, **kwargs)
+
+
+class ErrorNotification(Notification):
+    exception: BaseException
+
+    def __init__(self, exception: BaseException, *args, **kwargs):
+        msg = getattr(exception, 'message', str(exception))
+        severity = getattr(exception, 'severity', 'ERROR')
+        actions = getattr(exception, 'actions', ())
+        super().__init__(msg, severity, actions, type='error')
+        self.exception = exception
+
+
+class WarningNotification(Notification):
+    warning: Warning
+
+    def __init__(self, warning: Warning, *args, **kwargs):
+        msg = getattr(warning, 'message', str(warning))
+        severity = getattr(warning, 'severity', 'WARNING')
+        actions = getattr(warning, 'actions', ())
+        super().__init__(msg, severity, actions, type='warning')
+        self.warning = warning
+
+
+class NotificationManager:
+    """
+    A Singleton instance of notification manager.
+
+    Only one instance can be constructed; as we need notification to all flow to
+    a single location that is registered with the sys.except_hook  and
+    showwarning hook.
+
+    This can and should be used a context manager; the context manager will
+    properly rentered, and install/remove hooks and keep them in a stack to
+    restore them.
+
+    While it might seem unnecessary to make it entrant; or to make the
+    re-entrancy no-op; one need to consider that this could be used inside
+    another context manager that modify except_hook and showwarning.
+
+    Currently the original except and show warnings hooks are not called; but
+    this could be changed in the future; this poses some questions with the
+    re-entrency of the hooks themselves.
+    """
+
+    records: List[Notification]
+    _instance: Optional[NotificationManager] = None
+
+    def __new__(cls):
+        # singleton
+        if cls._instance is None:
+            cls._instance = object.__new__(cls)
+        else:
+            raise RuntimeError(
+                'Napari expects a single instance of NotificationManager'
+            )
+        return cls._instance
```

Comment:
I don't understand this logic... Shouldn't we just always return cls._instance, whether we had to create it or not?

Reply:
No we want to avoid creating many; tally originally went with a singleton, so indeed returning cls._instance if we try to create a new one; but from discussions and my personal experience Singletons area a difficult to test pattern, and impossible to remove or change later if we decide we do want multiple instance.

Reply:
I’m a little confused.  Isn’t this still a singleton?  As mentioned earlier, my original implementation was just a first draft.  If you want to change it that’s definitely ok with me